File: frontend/src/readiness.ts
export type ServiceReadiness = {
  configured: boolean;
  label: string;
  detail: string;
};

export type ReadinessResponse = {
  services: Record<string, ServiceReadiness>;
};

export type ReadinessSummary = {
  configuredCount: number;
  totalCount: number;
  tone: "ready" | "partial" | "empty";
  headline: string;
};

export function buildReadinessSummary(readiness: ReadinessResponse | null): ReadinessSummary {
  const services = Object.values(readiness?.services ?? {});
  const totalCount = services.length;
  const configuredCount = services.filter((service) => service.configured).length;

  if (totalCount === 0) {
    return {
      configuredCount: 0,
      totalCount: 0,
      tone: "empty",
      headline: "正在检查演示链路",
    };
  }

  if (configuredCount === totalCount) {
    return {
      configuredCount,
      totalCount,
      tone: "ready",
      headline: "演示链路已就绪",
    };
  }

  return {
    configuredCount,
    totalCount,
    tone: "partial",
    headline: "部分链路使用兜底",
  };
}

export function orderedReadinessServices(readiness: ReadinessResponse | null) {
  const services = readiness?.services ?? {};
  return [
    services.openai_realtime,
    services.gemini_live,
    services.azure_pronunciation,
    services.report_llm,
  ].filter((service): service is ServiceReadiness => Boolean(service));
}

File: backend/app/routes/readiness.py
from fastapi import APIRouter, Depends

from app.config import Settings
from app.dependencies import get_app_settings
from app.schemas import ReadinessResponse


router = APIRouter(prefix="/api", tags=["readiness"])


def _service(configured: bool, label: str, ready_detail: str, missing_detail: str) -> dict[str, object]:
    return {
        "configured": configured,
        "label": label,
        "detail": ready_detail if configured else missing_detail,
    }


@router.get("/readiness", response_model=ReadinessResponse)
def get_readiness(settings: Settings = Depends(get_app_settings)) -> ReadinessResponse:
    report_provider = settings.report_llm_provider.strip().lower()
    report_ready = (
        (report_provider == "gemini" and bool(settings.gemini_api_key))
        or (report_provider == "deepseek" and bool(settings.deepseek_api_key))
        or report_provider == "rules"
    )
    report_label = {
        "gemini": "Gemini report",
        "deepseek": "DeepSeek report",
        "rules": "Rules fallback",
    }.get(report_provider, "Report LLM")

    return ReadinessResponse.model_validate(
        {
            "services": {
                "openai_realtime": _service(
                    bool(settings.openai_api_key),
                    "OpenAI Realtime",
                    "Realtime voice is configured.",
                    "OpenAI credential is missing; use Gemini Live or demo turns.",
                ),
                "gemini_live": _service(
                    bool(settings.gemini_api_key),
                    "Gemini Live",
                    "Gemini Live token flow is configured.",
                    "Gemini credential is missing; OpenAI voice and demo turns can still run.",
                ),
                "azure_pronunciation": _service(
                    bool(settings.azure_speech_key and settings.azure_speech_region),
                    "Azure pronunciation",
                    "Pronunciation assessment is configured.",
                    "Azure Speech credential or region is missing; drills will show setup guidance.",
                ),
                "report_llm": _service(
                    report_ready,
                    report_label,
                    "Report enhancement is configured.",
                    "Report LLM is not configured; rules fallback will generate reports.",
                ),
            }
        }
    )
